### PR TITLE
gl_engine: dynamic load GL functions with glad

### PR DESCRIPTION
**Pull request by RuiwenTang**
This PR shows how to dynamic load GL functions with [GLAD](https://github.com/dav1dde/glad-web) (one of the [OpenGL Loader Library](https://www.khronos.org/opengl/wiki/OpenGL_Loading_Library)).

I choose `GLAD`  because it only contains a single source file, and does not need to link OpenGL library at compile time.

Issue: https://github.com/thorvg/thorvg/issues/2453

### CODE REVIEW

**Comment by vtorri** [comment]
on Windows, in my fork, i've updated the code, applied the PR and compiled with this script:

```
#! /bin/sh

rm -rf builddir

meson setup \
      --buildtype=debug \
      -Dprefix=/opt/thorvg_64 \
      -Dengines=sw,gl \
      -Dloaders=all \
      -Dsavers=all \
      -Dbindings=capi \
      -Dtools=all \
      -Dexamples=true \
      -Dtests=false \
      builddir
ninja -C builddir install
```

both `Lottie gl` and `Performance gl` are segfaulting. Without `gl`, software is working

gdb reports nothing usefull:

```
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007ffc8ff6b8de in libthorvg-0!_ZN3tvg8WgCanvas3genEv () from C:\Documents\msys2_2\opt\thorvg_64\bin\libthorvg-0.dll
#2  0x00007ffc90005928 in ?? () from C:\Documents\msys2_2\opt\thorvg_64\bin\libthorvg-0.dll
#3  0x00007ffc90006178 in ?? () from C:\Documents\msys2_2\opt\thorvg_64\bin\libthorvg-0.dll
#4  0x0000000000000000 in ?? ()
```

**Comment by RuiwenTang** [comment]
> on Windows, in my fork, i've updated the code, applied the PR and compiled with this script:
> 
> ```
> #! /bin/sh
> 
> rm -rf builddir
> 
> meson setup \
>       --buildtype=debug \
>       -Dprefix=/opt/thorvg_64 \
>       -Dengines=sw,gl \
>       -Dloaders=all \
>       -Dsavers=all \
>       -Dbindings=capi \
>       -Dtools=all \
>       -Dexamples=true \
>       -Dtests=false \
>       builddir
> ninja -C builddir install
> ```
> 
> both `Lottie gl` and `Performance gl` are segfaulting. Without `gl`, software is working
> 
> gdb reports nothing usefull:
> 
> ```
> (gdb) bt
> #0  0x0000000000000000 in ?? ()
> #1  0x00007ffc8ff6b8de in libthorvg-0!_ZN3tvg8WgCanvas3genEv () from C:\Documents\msys2_2\opt\thorvg_64\bin\libthorvg-0.dll
> #2  0x00007ffc90005928 in ?? () from C:\Documents\msys2_2\opt\thorvg_64\bin\libthorvg-0.dll
> #3  0x00007ffc90006178 in ?? () from C:\Documents\msys2_2\opt\thorvg_64\bin\libthorvg-0.dll
> #4  0x0000000000000000 in ?? ()
> ```


`libthorvg-0!_ZN3tvg8WgCanvas3genEv` this seems the code jump into WebGPU backend ?

**Comment by vtorri** [comment]
wg engine is disabled:

```
$ ls builddir/src/renderer/
gl_engine/  sw_engine/
```

**Comment by vtorri** [comment]
ok, I understand why gdb is reporting nothing useful: you strip by default. Here the backtrace

```
#0  0x0000000000000000 in ?? ()
#1  0x00007ffc84b5f5e4 in GlComposeTask::run (this=0x5761cf0) at ../src/renderer/gl_engine/tvgGlRenderTask.cpp:190
#2  0x00007ffc84b5fa01 in GlBlitTask::run (this=0x5761cf0) at ../src/renderer/gl_engine/tvgGlRenderTask.cpp:230
#3  0x00007ffc84b5ba65 in GlRenderer::sync (this=0x6bc7f0) at ../src/renderer/gl_engine/tvgGlRenderer.cpp:799
#4  0x00007ffc84c39c7d in tvg::Canvas::Impl::sync (this=0x6bc950) at ../src/renderer/tvgCanvas.h:127
#5  0x00007ffc84b31bea in tvg::Canvas::sync (this=0x6bc7a0) at ../src/renderer/tvgCanvas.cpp:86
#6  0x00007ff712c49446 in tvgexam::Window::ready (this=0x6b6500) at ../examples/Example.h:184
#7  0x00007ff712c418c8 in tvgexam::SDL_main (example=0x6b64b0, argc=2, argv=0x7a4ac0, width=1024, height=1024, threadsCnt=4, print=true)
    at ../examples/Example.h:522
#8  0x00007ff712c41950 in SDL_main (argc=2, argv=0x7a4ac0) at ../examples/Performance.cpp:80
#9  0x00007ff712c42c0e in main_getcmdline ()
#10 0x00007ff712c412ee in __tmainCRTStartup () at C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:267
#11 0x00007ff712c41406 in mainCRTStartup () at C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:188
(gdb)
```

note that `glClearDepthf` is supported since opengl 4.1: https://registry.khronos.org/OpenGL-Refpages/gl4/html/glClearDepth.xhtml

**Comment by RuiwenTang** [comment]
Oh, looks we should to use `glClearDepth ` instead

**Comment by vtorri** [comment]
i confirm that there is no segfault with `glClearDepth`
the result is similar than with glew (it is normal) on Windows:
* `Lottie gl`: the icons are displayed, then immediatly after, the window becomes and remains black
* `Performance gl`: the tiger is flickering

**Comment by RuiwenTang** [comment]
Hmm, I think the flickering is caused by some bug inside GLEngine and not by the loader

**Comment by vtorri** [comment]
@RuiwenTang do you have a Windows computer to test it ?

**Comment by RuiwenTang** [comment]
> @RuiwenTang do you have a Windows computer to test it ?

No , I don't have Windows development environment :(

**Comment by vtorri** [comment]
you may try on linux, with wine + cross compilation, to see if there is the same problem

**Comment by hermet** [comment]
> tw is there a reason to stick to OpenGL 3.3 and not upgrading to 4.1 ? OpenGL 4.1 was released in 2010, so it's quite old, no ? In addition, you could use the tesselation available in OpenGL 4.0.

@vtorri @RuiwenTang In the case of OpenGL, I think we can bump up the version if necessary.

**Comment by vtorri** [comment]
any news on this ?

**Comment by hermet** [comment]
@RuiwenTang looks busy. :) 
@vtorri Are there any issues on your side with using v3.3? Replacing tessellation shaders is another big task.

**Comment by vtorri** [comment]
@hermet except what i have said in https://github.com/thorvg/thorvg/pull/2574#issuecomment-2235241224 , no other issue

**Comment by RuiwenTang** [comment]
Sorry, I've been very busy lately and don't have time to continue with this matter :(

**Comment by lmdsp** [comment]
For the record, macos only supports OpenGL 3.2 and 4.1 core profiles, but not 3.3.
https://developer.apple.com/documentation/appkit/1436146-opengl_profiles?language=objc
Also various sources report that 4.1 support is quite buggy, so 3.2 would be the safest choice here.

**Comment by hermet** [comment]
> For the record, macos only supports OpenGL 3.2 and 4.1 core profiles, but not 3.3.

To clarify, the ThorVG GL engine shouldn't target macOS, as OpenGL has been deprecated regardless of the version.

**Comment by RuiwenTang** [comment]
Merged the khr header and GL header into a single file, does this looks better ?

**Comment by hermet** [comment]
@RuiwenTang Well... for now, I don't think so.

**Comment by hermet** [comment]
> OpenGL 3.2 and 4.1 core profiles,

@RuiwenTang what do you think downgrade to 3.2?

**Comment by RuiwenTang** [comment]
> > OpenGL 3.2 and 4.1 core profiles,
> 
> @RuiwenTang what do you think downgrade to 3.2?

If macOS is not our target platform, I don't think it's necessary to downgrade the GL version for it, instead we should increase the minimum version requirement to use the new GL features in other platform.

**Comment by lmdsp** [comment]
> > > OpenGL 3.2 and 4.1 core profiles,
> > 
> > 
> > @RuiwenTang what do you think downgrade to 3.2?
> 
> If macOS is not our target platform, I don't think it's necessary to downgrade the GL version for it, instead we should increase the minimum version requirement to use the new GL features in other platform.

Just for the record, although OpenGL has been deprecated for quite a while on macOS, it has still been supported and been working great since then. There's a myriad of applications in fields such as science, audio, 2D/3D graphics etc which rely on OpenGL and Apple probably wants to ensure these still work.

**Comment by RuiwenTang** [comment]
> @RuiwenTang Care to weigh in? I don't see any ThorVG issues on macOS. If the downgrade is (almost) costless, I think it's good to go.

The shader is written for GLSL3.30 which can defined locations for [attributes](https://www.khronos.org/opengl/wiki/Layout_Vertex_Attribute) and [fragment shader outputs](https://www.khronos.org/opengl/wiki/Layout_Fragment_Output) in shader directly. If downgrade to GL 3.2 , all shaders and associated rendering codes need to rewrite. 
see the reference : https://www.khronos.org/opengl/wiki/History_of_OpenGL

**Comment by hermet** [comment]
@RuiwenTang I'm a bit confused. As @lmdsp mentioned, macOS only supports v3.2. How is ThorVG running with v3.3 on macOS now?

**Comment by RuiwenTang** [comment]
> @RuiwenTang I'm a bit confused. As @lmdsp mentioned, macOS only supports v3.2. How is ThorVG running with v3.3 on macOS now?

According to this: http://support.apple.com/kb/HT5942
Most macbook support OpenGL 4.1, some old hardware support OpenGL 3.3

**Comment by vtorri** [comment]
any news about this PR ?

**Comment by RuiwenTang** [comment]
> any news about this PR ?

Not yet. If glad can't meet the requirements, it seems that we need to write a loader by our own. :(

**Comment by hermet** [comment]
@RuiwenTang I'm waiting for the DRAFT tag off and the Android build fix.

**Comment by vtorri** [comment]
@RuiwenTang no idea why android builds failed  in CI ?

File: src/renderer/gl_engine/meson.build [hunk 1]
```diff
@@ -51,9 +53,10 @@ else
     gl_target_profile = '-DTHORVG_GL_TARGET_GL=1'
 endif
 
+add_languages('c', required: true)
```
Comment (hermet) [inline src/renderer/gl_engine/meson.build:56]: image

due to this, android has been failed, we don't need to use c language here?

**Review by hermet — COMMENTED** [review]
@RuiwenTang please check a comment.

**Comment by RuiwenTang** [comment]
Can we move to new PR #2866 ？which use the branch in the main repositry

**Pull request closed without merging**